### PR TITLE
Prob instead of exp val

### LINKED ISSUE #16: Change invariant QNN to be equivariant
To achieve scalar-invariance in $`E_{xc}^{\theta}[\rho]`$, the coefficient function $`c_{\theta}[\rho](r_i)`$ must be equivariant.

When working with amplitude embedding, equivariance may be enforced by associating projector measurements with grid points in a particular way. For example, take a density evaluated at four points, flattened into a 1d vector $\rho = [\rho(r_1), \rho(r_2), \rho(r_3), \rho(r_4)]$. Embedding into a two qubit quantum state with amplitude embedding (ignoring the normalization factor) we have $|\rho \rangle = \rho(r_1)|00\rangle +\rho(r_2)|01\rangle + +\rho(r_3)|10\rangle + +\rho(r_4)|11\rangle$. Now, we associate each grid point with $r_i$ by using the projector corresponding to the basis state that $\rho(r_i)$ multiplies. In this case, $r_1 = |00 \rangle \langle 00|$, $r_2 = |01 \rangle \langle 01|$, $r_3 = |10 \rangle \langle 10|$, $r_4 = |11 \rangle \langle 11|$.

This guarantees equivariance in $`c_{\theta}[\rho](r_i)`$ as detailed in the paper and reduces it to $`c_{\boldsymbol{\theta}}[\rho](\mathbf{r}_i) = \bigl|\langle i|\Psi(\rho,\boldsymbol{\theta})\rangle\bigr|^{2}`$ which is simply the probability of measuring the bitstring $i$, obtained easily from `qml.probs` (https://docs.pennylane.ai/en/stable/code/api/pennylane.probs.html).

We should change our measurements to adhere to this new rules.

### PR DESCRIPTION
**Pull request by minhtriet**
fix #16

### CODE REVIEW

File: tests/test_circuit_equivariance.py [hunk 1]
```diff
@@ -88,16 +80,17 @@ def test_a_training_step_6qb_d4():
...
+    assert np.allclose(result_rot_y, _180_y_rot @ result)
+    assert np.allclose(result_rot_z, _180_z_rot @ result)
+    assert np.allclose(result_rot_mock_coeff_inputs_x_y_eq_z, _270_x_y_eq_z @ result)
 
     densities = normal(key, shape=(2**num_wires, 1))
 
     result = result[:, jnp.newaxis]
     result_rot_x = result_rot_x[:, jnp.newaxis]
 
+    # @jack this fail, do we need to rotate result?
```
Comment (minhtriet) [inline tests/test_circuit_equivariance.py:93]: @jackbaker1001 any idea here? By skimming your explanation perhaps we don't need to do that?

Comment (jackbaker1001) [reply tests/test_circuit_equivariance.py:93]:
```python
    densities = normal(key, shape=(2**num_wires, 1))

    result = result[:, jnp.newaxis]
    result_rot_x = result_rot_x[:, jnp.newaxis]

    # @jack this fail, do we need to rotate result?
    xc_energy_density = jnp.einsum("rf,rf->r", result, densities)
    xc_energy_density = abs_clip(xc_energy_density, 1e-30)

    xc_energy_density_rot_x = jnp.einsum("rf,rf->r", result_rot_x, densities)
    xc_energy_density_rot_x = abs_clip(xc_energy_density_rot_x, 1e-30)

    grid_weights = normal(key, shape=(2**num_wires,))
    assert np.isclose(
        _integrate(xc_energy_density, grid_weights),
        _integrate(xc_energy_density_rot_x, grid_weights),
    )
```

You've confused yourself a little here, and i'm not entirely sure what you are trying to test.

if you are trying to test for invariance in E_xc, you must do the integral with inputs: coefficient function(rotated input) and energy density (rotated input) and compare this with  coefficient function(unrotated input) and energy density (unrotated input).

For the energy density part, functionals which compute this are naturally equivariant. For this mock case, the rotated and unrotated energy density can just be obtained by picking some random energy density (`densities` in your code) and rotating it using the rotation matrix `Ug_mat = qml.matrix(_180_x_rot[0])`

Also, the grid weights cannot be random for this to work. Each grid weight must be identical, as mentioned in the paper.

Comment (jackbaker1001) [reply tests/test_circuit_equivariance.py:93]: The rest of this PR does look correct though and mirrors what I have also independently done. Good job.

Comment (jackbaker1001) [reply tests/test_circuit_equivariance.py:93]: Basically @minhtriet , I think you need to rotate `densities` using `Ug_mat` here:

`xc_energy_density_rot_x = jnp.einsum("rf,rf->r", result_rot_x, densities)`

changes to

`xc_energy_density_rot_x = jnp.einsum("rf,rf->r", result_rot_x, Ug_mat @ densities)`

and generate uniform integration grid weights.

**Review by jackbaker1001 — CHANGES_REQUESTED** [review]
See conversation above for changes.

**Review by jackbaker1001 — APPROVED** [review]
LGTM now. Will merge.
